source/dem/particle_wall_contact_force.cc

```diff
@@ -158,209 +165,445 @@ ParticleWallContactForce<dim,
-                            tangential_force,
-                            tangential_torque,
-                            rolling_resistance_torque);
-
-                          // Applying the calculated forces and torques on the
-                          // particle
-                          types::particle_index particle_id =
-                            particle->get_local_index();
-                          Tensor<1, 3> &particle_torque =
-                            contact_outcome.torque[particle_id];
-                          Tensor<1, 3> &particle_force =
-                            contact_outcome.force[particle_id];
-
-                          this->apply_force_and_torque(
-                            normal_force,
-                            tangential_force,
-                            tangential_torque,
-                            rolling_resistance_torque,
-                            particle_torque,
-                            particle_force);
+                          contact_record[particle->get_local_index()]
+                            .emplace_back(triangle_cell_iterator,
+                                          normal_overlap,
+                                          contact_indicators[particle_counter],
+                                          contact_info);
                         }
-                      else
+                    }
+                  particle_counter++;
+                }
+            }
+        }
+      // Every contact between this solid object and the particles are
+      // recorded. Now, we need to remove the invalid contacts from
+      // the records and compute the forces associated with the valid
+      // contacts.
+
+      // Loop on every particle / contact record
+      for (auto &[particle_index, this_contact_record] : contact_record)
+        {
+          // Loop on every contact in the contact record
+          for (auto contact_1 = this_contact_record.begin();
+               contact_1 != this_contact_record.end();)
+            {
+              // Extract the information of C1
+              auto &[T1_cell,
+                     normal_overlap_C1,
+                     contact_indicator_C1,
+                     contact_info_C1] = *contact_1;
+
+              // Assigning the triangle neighboring list of T1;
+              const std::vector<
+                typename Triangulation<dim - 1, dim>::active_cell_iterator>
+                &T1_es_neighbors = this_solid_es_neighbors.at(T1_cell);
+              const std::vector<
+                typename Triangulation<dim - 1, dim>::active_cell_iterator>
+                &T1_vs_neighbors = this_solid_vs_neighbors.at(T1_cell);
+
+              // Initialize variables. We need to compare C1 to every other
+              // contact that was compared yet.
+              bool erase_contact_1 = false;
+              auto contact_2       = std::next(contact_1);
+              while (contact_2 != this_contact_record.end())
```

Comment:
Instead of calling them _1 and _2, I would try to use a more descriptive name like contact_with_triangle_2 or contact_triangle_2. Something that is a bit more descriptive than just 1 and 2. Especially since you have C1 and C2 after that it makes it a bit confusing.